source/compilation/Compilation.cpp

```diff
@@ -458,19 +458,23 @@ void Compilation::createDefinition(const Scope& scope, LookupLocation location,
     // Record that the given scope contains this definition. If the scope is a compilation unit, add
     // it to the root scope instead so that lookups from other compilation units will find it.
     auto targetScope = scope.asSymbol().kind == SymbolKind::CompilationUnit ? root.get() : &scope;
+    bool redefined = false;
     std::tuple key(def->name, targetScope);
     if (auto it = definitionMap.find(key); it != definitionMap.end()) {
         reportRedefinition(scope, *def, *it->second);
-        return;
+        redefined = true;
     }
 
-    auto [it, inserted] = definitionMap.emplace(key, std::move(def));
-    ASSERT(inserted);
+    auto [it, inserted] = definitionMap.insert_or_assign(key, std::move(def));
+    if (!redefined) {
+        ASSERT(inserted);
+    }
 
     auto result = it->second.get();
     if (targetScope == root.get()) {
         auto& topDef = topDefinitions[result->name].first;
```

Comment:
Same here -- I would just rewrite this as `topDefinitions[result->name].first = result;` now that it's allowed for there to be duplicates.

Reply:
I was trying to replace:
```
    auto [it, inserted] = definitionMap.insert_or_assign(key, std::move(def));
    //if (!redefined) {
    //    ASSERT(inserted);
    //}

    auto result = it->second.get();
    if (targetScope == root.get()) {
        topDefinitions[result->name].first = result;
```
with something simpler:
```
    const auto& result = (definitionMap[key] = std::move(def));
    if (targetScope == root.get()) {
        topDefinitions[result->name].first = &(*result);
```
It compiles cleanly, passes unittest, but the ugly `&(*result)` syntax, and the fact that `std::uniq_ptr` , x-values (`std::move`) 
 is pretty new to me, scares me a bit.
I switched today to Ubuntu 22.04 (from 20.04) and clang-14 (from g++-9) since I never got g++ failures without clang, but got clang failures when g++ passed.

Reply:
You should be able to do a `result.get()` instead in your example, which is a bit nicer. I think you still want the insert_or_assign though, and then use the `inserted` result to run your `reportRedefinition`. Then you can get rid of the extra `find()` call.

Reply:
Hmm actually, that suggestion to collapse the find() call doesn't work because we do want to see the old element for purposes of error / warning reporting before we overwrite it.